config/bash/lxc.in

```diff
@@ -1,111 +1,1031 @@
+            ;;
+        --keep-var )
+            COMPREPLY=( $( compgen -A variable -- "$cur" ) )
+            return
+            ;;
+        --uid | -u )
+            _uids
+            return
             ;;
-        esac
+        --gid | -g )
+            _gids
+            return
+            ;;
+        --context | -c )
+            # @TODO: list all SElinux contexts available.
+            return
+            ;;
+    esac
 
-        return 1
-    }
+    $split && return
+
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_attach lxc-attach
+
+__lxc_groups() {
+    declare -A groups
+    for line in "$(command lxc-ls -f --fancy-format GROUPS | command sed -e '/^-/d' -e '1d' -e 's/,/ /g')"; do
+	for grp in $line; do
+            groups+=([${grp}]=1)
+        done
+    done
+    printf "%s " "${!groups[@]}"
+}
 
-    _lxc_generic_ns() {
-        local cur prev
+_lxc_autostart() {
+    local cur prev words cword split
+    COMPREPLY=()
+    _init_completion -s -n : || return
 
-        COMPREPLY=()
-        _get_comp_words_by_ref cur prev
+    _lxc_common_opt || return
 
-        case $prev in
-            -n)
-                _lxc_names "$cur"
-                return 0
+    case $prev in
+        --reboot | -r | --shutdown | -s | --kill | -k | --list | -L | --all | -a | --ignore-auto | -A )
+            # Only flags.
+            ;;
+        --timeout | -t )
+            COMPREPLY=( $( compgen -P "$cur" -W "{0..9}" ) )
+            compopt -o nospace
+            return
             ;;
+        --groups | -g )
+            # @TODO: add NULL group as a leading comma, trailing comma, embedded double comma.
+            __lxc_piped_args "$cur" ',' $( __lxc_groups )
+            compopt -o nospace
+            return
+            ;;
+    esac
+
+    $split && return
+
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+}
+complete -F _lxc_autostart lxc-autostart
 
-            -s)
-                _lxc_states "$cur"
-                return 0
+_lxc_cgroup() {
+    local cur prev words cword split
+    COMPREPLY=()
+    _init_completion -s -n : || return
+
+    _lxc_common_opt || return
+
+    case $prev in
+        --name | -n )
+            _lxc_names
+            return
             ;;
-        esac
+        --rcfile )
+            _filedir
+            return
+            ;;
+    esac
 
-        return 1
-    }
+    $split && return
 
-    _lxc_generic_t() {
-        local cur prev
+    if [[ $cur == -* ]]; then
+        COMPREPLY=( $( compgen -W '$( _parse_help "$1" )' -- "$cur" ) )
+        [[ ${COMPREPLY-} == *= ]] && compopt -o nospace
+        return
+    fi
+    _lxc_append_name
+}
+complete -F _lxc_cgroup lxc-cgroup
```

Comment:
For `lxc-cgroup`, the last params are the `state-object` and its new value (optional). How feasible is something like:
```
__lxc_list_cgroup_state_object() {
    local userSlicePath="/sys/fs/cgroup/user.slice"
    if [[ -d "${userSlicePath}" ]] && [[ -f "${userSlicePath}/cgroup.controllers" ]]; then
        for cont in $(<"${userSlicePath}/cgroup.controllers"); do
            for entry in ${userSlicePath}/${cont}.*; do
                [[ ! -f "${entry}" ]] && continue
                local param="${entry##*/}"
                COMPREPLY+=( $( compgen -W "${param}" -- "$cur" ) )
            done
        done
    fi
}
```

to list the available values? Considering we don't want to use external tools.